source/grid/grid_tools.cc

```diff
@@ -4352,284 +4351,102 @@ namespace GridTools
-                found              = true;
-                cell_candidate_idx = box_cell.size();
-                box_cell.push_back(
-                  std::make_pair(my_pair.first->bounding_box(), my_pair.first));
-              }
-            catch (const GridTools::ExcPointNotFound<dim> &)
-              {
-                missing_points_out.emplace_back(points_checked);
-              }
+            cells_out.emplace_back(cell);
+            qpoints_out.resize(cells_out.size());
+            maps_out.resize(cells_out.size());
+            return (cells_out.size() - 1);
           }
+      };
 
-        // Updating the position of the analyzed points
-        ++points_checked;
-      }
-
-    // If the point has been found in a cell, adding it
-    if (found)
-      {
-        cells_out.emplace_back(my_pair.first);
-        qpoints_out.emplace_back(1, my_pair.second);
-        maps_out.emplace_back(1, points_checked - 1);
-      }
-
-    // Now the second easy case.
-    if (np == qpoints_out.size())
-      return std::make_tuple(std::move(cells_out),
-                             std::move(qpoints_out),
-                             std::move(maps_out),
-                             std::move(missing_points_out));
-
-    // Cycle over all points left
-    for (unsigned int p = points_checked; p < np; ++p)
-      {
-        // We assume the last used cell contains the point: checking it
-        if (cell_candidate_idx != -1)
-          if (!box_cell[cell_candidate_idx].first.point_inside(points[p]))
-            // Point outside candidate cell: we have no candidate
-            cell_candidate_idx = -1;
-
-        // If there's no candidate, run a tree search
-        if (cell_candidate_idx == -1)
-          {
-            // Using the b_tree to find new candidates
-            box_cell.clear();
-            b_tree.query(boost::geometry::index::intersects(points[p]),
-                         std::back_inserter(box_cell));
-            // Checking the returned bounding boxes/cells
-            use_try            = false;
-            cell_candidate_idx = -1;
-            for (unsigned int i = 0; i < box_cell.size(); ++i)
-              {
-                // As a candidate we don't want artificial cells
-                if (!box_cell[i].second->is_artificial())
-                  cell_candidate_idx = i;
-
-                // If the cell is not locally owned or at boundary
-                // we check for exceptions
-                if (cell_candidate_idx != -1 &&
-                    (!box_cell[i].second->is_locally_owned() ||
-                     box_cell[i].second->at_boundary()))
-                  use_try = true;
-
-                // If a cell candidate was found we can stop
-                if (cell_candidate_idx != -1)
-                  break;
-              }
-          }
+    // Check all points within a given pair of box and cell
+    const auto check_all_points_within_box =
+      [&](const decltype(*b_tree.begin()) &leaf) {
```

Comment:
since you're already using a generic lambda with the second and third argument, why can't you just use
```suggestion
      [&](const auto &leaf) {
```
here?